### ISSUE TITLE
How to play video files that require permission?

### ISSUE THREAD

**Issue by bitdessert** [mattermost/mattermost-mobile#5320]
Submit feature requests to http://www.mattermost.org/feature-requests/. File non-security related bugs here in the following format: 

#### Summary
How to play video files that require permission?

#### Environment Information

- Device Name:  iPhone 12
- OS Version: 
- Mattermost App Version: 1.41.1 
- Mattermost Server Version:5.32.1

#### Steps to reproduce
Upload a video file to the channel.
Click on the file.
File cannot be played

#### Expected behavior
If the user is already logged in and the user has uploaded the video file, it should be played.

#### Observed behavior
File cannot be played

#### Possible fixes
When I check, the file url is received well.
However, the url is a url that cannot be accessed without permission.
I tried to play with the link published through 'Client4.getFilePublicLink', but it does not play.

**Comment by amyblais** [comment]
@josephbaylon Do you know?

**Comment by josephbaylon** [comment]
@amyblais @bitdessert I'm a little unsure about this. Is this about uploading an actual video from device or posting a video url on the channel. If the latter, then it needs to be public.

**Comment by bitdessert** [comment]
@amyblais @josephbaylon Thank you for your answer.
Prior to version 1.36.0, files were opened after downloading.
When the user clicks download and the download is complete, the file is played.

It seems that a lot of parts have changed as the gallery component was changed from 1.41.1.
In the current version, the file is not downloaded and it tries to play it with only the URL.
Click Share, and after the file is saved in the cache folder, the file is played.

It seems that you need some logic to save the file before playing it.
In the case of a video file, it is a logic that automatically saves it first.

**Comment by josephbaylon** [comment]
@bitdessert To download a file attachment,
1. Tap on attachment
2. Tap share button at the bottom right

In case of iOS, usually there is a **Save to Files** option, otherwise you can share through other apps that lets you download the file to the device.

In case of Android, usually if you don't see the save to device option, which is dependent on your android device, you can share via google drive and download from there.

For video attachments, the following are the supported file types that will display the playback button, but it's not always guaranteed to have the playback button.

iOS: mp4, x-m4v, quicktime
Android: 3gpp, x-matroska, mp4, webm

If Download button appears instead of playback, tapping on Download button is the same as tapping on share button.

Hope this clarifies the gallery functionality.

**Comment by bitdessert** [comment]
thanks, @josephbaylon 

The problem now is that after tapping the uploaded video file, it won't play. 

If you tap on a video file that has not been downloaded for the first time, an error message pops up.

In 'gallery_video.tsx' file
'mobile.video_playback.failed_description'
An error occurred while trying to play the video.

**Comment by josephbaylon** [comment]
@bitdessert  Mind sharing the video in question?

**Comment by josephbaylon** [comment]
@bitdessert Can you try this locally?
1. On a channel, tap on capture video icon
2. Create and use the video
3. Upload the video
4. Tap on the video you just uploaded

**Comment by bitdessert** [comment]
@josephbaylon  It's not a video file problem.

The problem is as follows. 
1. Upload the file. 
2. Tap the file.

When you tap the file, the file url is fetched from the source code below.
'setUri(exist ? localPath : Client4.getFileUrl(file.id, Date.now()));'
At this time, the url I got is not a public URL. 
When accessing the URL, json is returned. 
The json content is that I don't have permission to access the file.
In the previous version, I could download the video file before playing it. 
And because the downloaded file is stored in the cache folder, it was possible to play the file without permission problem.

Now, when I try to play the video by directly accessing the URL before the file is downloaded, a problem occurs.

So, it seems necessary to download the file before playing it.
In other words, even if you do not have the file permission, because the file has been downloaded, it is possible to play it.

Even in the current version, after an error pop-up occurs, download the file anyway and tap it again to play normally.

**Comment by josephbaylon** [comment]
@enahum I don't think I'm understanding the issue, if there's any. Can you please chime in on this and see if there's a current limitation/restriction or the behavior is as expected?

**Comment by bitdessert** [comment]
https://user-images.githubusercontent.com/42505950/114803315-31457900-9dda-11eb-820d-47c28802e23c.mov

**Comment by bitdessert** [comment]
Url when first tapped : https://~~~/api/v4/files/zkpt5f1rhpr7bbb31r64na9aew?1618452093697
The above url does not have access rights.

The file does not have access rights, but the downloaded file is played.

After tapping the share button, url : /~~~/CoreSimulator/Devices/6A75B1D9-D85C-4C8D-ACB9-83ECEF10155C/data/Containers/Data/Application/1A922E7C-398D-4862-A1DC-9DB9C1997583/Library/Caches/Videos/video_progressbar.mov

**Comment by enahum** [comment]
This doesn't look like a permission issue but more of a streaming problem.

**Comment by bitdessert** [comment]
@enahum 
When the above address is accessed from a browser, a json return as shown below is received.
https://~~~/api/v4/files/zkpt5f1rhpr7bbb31r64na9aew?1618452093697
![image](https://user-images.githubusercontent.com/42505950/114803825-293a0900-9ddb-11eb-8939-c2e43ef43d51.png)

**Comment by enahum** [comment]
If you access the same url from a browser but after logging in into MM you'll get a different result

**Comment by bitdessert** [comment]
@enahum 
I just tried accessing that URL in a browser. 
I am logged in on mobile.

https://user-images.githubusercontent.com/42505950/114804180-da40a380-9ddb-11eb-8d4d-6e0a66e49a89.mov

**Comment by bitdessert** [comment]
@enahum 
I think that the module called 'react-native-video' tries to play the video by accessing the url in the get method.
The file should be played after checking the authority by sending the login information in the header together, 
but I think that an error occurs because the login information is not sent.
That's why I suggested a way to download the file in advance. 
Of course, if you can send login information through react-native-video module, it would be nice to do so.

**Comment by bitdessert** [comment]
@enahum @josephbaylon 
The video seemed to be confusing, so I recorded it again.


https://user-images.githubusercontent.com/42505950/114817330-c2c1e480-9df4-11eb-9682-795338495659.mov

**Comment by bitdessert** [comment]
@enahum @josephbaylon 
The video uploaded as an attachment seems to have not been properly recorded, so I recorded it again and uploaded it.

https://user-images.githubusercontent.com/42505950/114819737-e0914880-9df8-11eb-98ab-7a36084e92f9.mov

**Comment by nerthazrim** [comment]
Exact same issue on Android 9, 10 or 11.

Note that I can playback a video I uploaded myself, though.

**Comment by larkox** [comment]
@bitdessert there has been a few changes in the gallery on V2. Can you check if this is still happening to you on the latest beta?
